state_machine_design.py

```diff
@@ -118,30 +118,45 @@ def process(self, sm):
         print("CHECK_TIME process func")
         print("ens index: " + str(sm.ens_index))
 
+        # this is a small window where if the scheduler wakes up
+        # slightly early from sleep we allow it to run the ensemble anyway
+        TIME_BUFFER = 5
+
+
         if sm.ens_index < len(sm.ens_list):
             # read time from ensemble and compare to current_time
             now = time.localtime()
-            curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+            curr_time_seconds = hms_to_seconds(now.tm_hour, now.tm_min, now.tm_sec)
 
             nearest_ens_time = sm.ens_list[sm.ens_index]["start_time"]
 
+                
             if nearest_ens_time < curr_time_seconds:
-                print("time past current ens, skipping")
-                self.check_time_ctrl = SKIP
-                sm.err_code = MISSED_ENS
-                # sm.err_msg = "Current ensemble past time, skipping ensemble "
-                # + sm.ens_list[sm.ens_index]["title"] + " at " + curr_time_seconds
-            elif nearest_ens_time == curr_time_seconds:
+                print("time past current ens, checking if should skip")
+                if sm.rst == True:
+                    if now.tm_wday == sm.day_of_ens:
```

Comment:
I might be missing something, but under what scenario does `sm.day_of_ens` not equal `now.tm_wday`? From my understanding, this is trying to test whether the next ensemble is supposed to execute the day after the current one. How would `sm.day_of_ens` be set to reflect this?

Reply:
This covers the situation where the final ensemble of day 1 begins before midnight but finishes after midnight. So the `sm.day_of_ens` would be day 1 and `now.tm_wday` would be day 2. The purpose is to differentiate finishing an ensemble during day 1 and the time being measured as later than the first ensemble of day 2, ie. finish at 10 pm and first ensemble of the day is 6 am (so the scheduler should go to sleep but the simple if statement says 10 pm is past 6 am and should be skipped) vs an ensemble finishing past the start time of the next ensemble, ie finish at 7 am and first ensemble of the day is 6 am (in which case the scheduler should skip the missed ensemble).

I think this scenario in actuality would be very rare (though I guess depends on when the ensembles are scheduled and how long they might actually take) but I tried to solve it just to have everything covered.

Reply:
Ah I see, thank you for explaining the approach.